File: 0377-combination-sum-iv/0377-combination-sum-iv.java
class Solution {
    public int combinationSum4(int[] nums, int target) {
        int[] dp = new int[target+1];
        Arrays.fill(dp, -1);
        return f(nums, dp, target);
    }
    int f(int[] nums, int[] dp, int target){
        if(target == 0) return 1;
        if(dp[target] != -1) return dp[target];
        dp[target] = 0;
        for(int x : nums){
            if(x <= target) dp[target] += f(nums, dp, target-x);
        }
        return dp[target];
    }
}

File: 0377-combination-sum-iv/0377-combination-sum-iv.py
class Solution:
    def combinationSum4(self, nums: List[int], target: int) -> int:
        dp = [-1] * (target+1)
        def f(target):
            if target==0: return 1
            if dp[target]!=-1: return dp[target]
            dp[target] = 0
            for x in nums:
                if x <= target:
                    dp[target] += f(target-x)
            return dp[target]
        return f(target)